### ISSUE TITLE
Using black code formatter for HyperSpy

### ISSUE THREAD

**Issue by tjof2** [hyperspy/hyperspy#2316]
Based on a discussion in #2312, I think its worth recording it in an issue here.

The proposal is to apply the [black](https://black.readthedocs.io/en/stable/) code formatter to HyperSpy at some point in the near future.

> @francisco-dlp 
> We've already passed the whole code through autopep8 once a while ago (just after the 1.0 release if I remember well). However, because not everybody uses autopep8 and we don't check for compliance in the PRs, the situation today is not as good as it should be as this PR shows.
> 
> However, I think this is not the time to run the code through autopep8 again, because it makes merging the old PRs a nightmare. Since we are planning to merge as many PRs as possible before the split, I think that just before the split will be a good time to do this.
> 
> Once we go for it, I am also in favor of using black.

Black can also be used in conjunction with pre-commit hooks, e.g. https://black.readthedocs.io/en/stable/version_control_integration.html

**Comment by tjof2** [comment]
Noting here also that [`pyupgrade`](https://github.com/asottile/pyupgrade) can be used for f-strings @ericpre.

Pandas and dask-ml are two big projects that have applied black to their codebase. Obviously reducing the open PRs to near zero would be optimal for minimizing conflicts, but maybe it is possible after some of the bigger PRs for 1.6 are done (e.g. non-uniform axes etc) because many of the other PRs look rather stalled.

**Comment by thomasaarholt** [comment]
How about we do this immediately after releasing 1.6? That way any formatting complications don't delay the release.

**Comment by dnjohnstone** [comment]
Are we going to do this pre or post 1.6 release @francisco-dlp and @tjof2 ? I'd be quite keen for this to be done for 1.6 but understand why that may not be wanted.

**Comment by francisco-dlp** [comment]
I agree that this would be highly beneficial, but I think that it is best to wait until "The Split" because it'll be a nightmare to solve the conflicts of the pending PRs (49 as of today) otherwise...

**Comment by dnjohnstone** [comment]
Ok, fair enough - great that the number of PRs is going in the right direction now though! :)

**Comment by thomasaarholt** [comment]
Summer is here and there are relatively few PRs that are fresh (and fewer than the 49 two years ago). It would be really nice if we could A) format the whole repo with black and B) add an auto-formatter of some kind.

@ericpre thoughts?

**Comment by magnunor** [comment]
I agree! Having a consistent code style is so nice, and it seem like most projects have converged on `black` as the default style formatter. I have a small preference for having an automatic style checker, which fails if any part of the code doesn't follow the black style checker. Instead of a automatic formatter. I feel the automatic formatters clutter the commit history.

Doing this is going to be painful no matter what, so I agree with doing it at a time where there are few outstanding pull requests.

We're also planning on getting a similar style checker in pyXem soonish.

**Comment by thomasaarholt** [comment]
I'm happy for there to be an automated style checker. We use `pre-commit` at work, and it's pretty neat once you get used to it.

**Comment by ericpre** [comment]
This sounds good.

> I'm happy for there to be an automated style checker. We use `pre-commit` at work, and it's pretty neat once you get used to it.

I don't have experience with pre-commit and the part which says "it's pretty neat once you get used to it" doesn't give me confidence that it would be suitable for irregular contributors!

**Comment by thomasaarholt** [comment]
Yeah, that is a good point. I was chewing on that myself as I was typing it. The challenge is how to encourage users to format their code without hurting new ones. Either you automatically format it for them, or you inform them in some manner. 

pre-commit does the latter by showing output like the following, with the option of automatically formatting it. The "get used to it part" is that you often commit, see that it failed, and then commit again after fixing. 
For hyperspy this would be less invasive than it is at my work, where docstrings are required and pre-commit often complains. If people use `format on save` in VSCode, then they should never have a problem with pre-commit.

![1_b1KAkP6CXPx669MAV3CoQg](https://user-images.githubusercontent.com/2721423/174888430-f40d6081-f5c9-44fe-b7e4-d918d2e85f49.png)
